File: LeetCode/Q24.c
/**
 * Given a linked list, swap every two adjacent nodes and return its head. 
 * You must solve the problem without modifying the values in the list's nodes 
 * (i.e., only nodes themselves may be changed.)
 * Scott 2021/08/20
*/
#include <stdlib.h>
#include <stdio.h>

struct ListNode {
    int val;
    struct ListNode *next;
};

struct ListNode *create_list(int *array, int size) {
    struct ListNode *node = malloc(sizeof(struct ListNode));
    struct ListNode *head = node;
    for (int i = 0; i < size - 1; ++i) {
        node->val = array[i];
        node->next = malloc(sizeof(struct ListNode));
        node = node->next;
    }
    node->val = array[size - 1];
    node->next = NULL;
    return head;
}

// Original
// 37.54%
// Same idea as the 100% one.
// But I failed to properly implement it.
struct ListNode* swapPairs1(struct ListNode* head) {
    struct ListNode *prev = NULL;
    struct ListNode *node = head;
    if (!node) {
        return NULL;
    }
    struct ListNode *next = node->next;
    if (!next) {
        return node;
    }
    struct ListNode *result = next;

    while (node && next) {
        node->next = next->next;
        next->next = node;
        
        if (prev) {
            prev->next = next;
        }

        prev = node;

        if (node->next && next->next) {
            struct ListNode *temp = node;
            node = next->next->next;
            next = temp->next->next;
        } else {
            break;
        }
    }

    return result;
}

int main(void) {

    int array[] = {1, 2, 3, 4, 5};
    struct ListNode *head = create_list(array, 5);
    struct ListNode *result = swapPairs1(head);
    while (result) {
        printf("%d->", result->val);
        result = result->next;
    }
    printf("NULL\n");

    return 0;
}

File: LeetCode/python/Q24.py
# 24. Swap Nodes in Pairs
# Medium
# Given a linked list, swap every two adjacent nodes and return its head.
# You must solve the problem without modifying the values in the list's nodes
# (i.e., only nodes themselves may be changed.)

from typing import Optional


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

    def create(l: list[int]):
        if len(l) == 0:
            return None
        head = ListNode(l[0])
        node = head
        for i in range(1, len(l)):
            node.next = ListNode(l[i])
            node = node.next
        return head

    def __repr__(self) -> str:
        node = self
        result = ''
        while node:
            result += str(node.val) + '->'
            node = node.next
        return result + 'None'


class Solution:

    # 2023-05-16 13:24:53
    # original
    # O(n)
    def swapPairs(self, head: Optional[ListNode]) -> Optional[ListNode]:
        curr = head
        head = ListNode(next=head)
        prev = head
        while curr is not None and curr.next is not None:
            next_ = curr.next
            curr.next, next_.next = next_.next, curr
            prev.next = next_
            prev = curr
            curr = curr.next
        return head.next


s = Solution()
print(s.swapPairs(ListNode.create([1, 2, 3, 4])))
print(s.swapPairs(ListNode.create([1, 2, 3, 4, 5, 6, 7, 8])))
print(s.swapPairs(ListNode.create([1, 2, 3, 4, 5])))
print(s.swapPairs(ListNode.create([1])))
print(s.swapPairs(None))
